packages/editor/src/lib/editor/derivations/notVisibleShapes.ts

```diff
@@ -0,0 +1,95 @@
+import { RESET_VALUE, computed, isUninitialized } from '@tldraw/state'
+import { TLPageId, TLShapeId, isShape, isShapeId } from '@tldraw/tlschema'
+import { Box } from '../../primitives/Box'
+import { Editor } from '../Editor'
+
+function isShapeNotVisible(editor: Editor, id: TLShapeId, viewportPageBounds: Box): boolean {
+	const maskedPageBounds = editor.getShapeMaskedPageBounds(id)
+	// if the shape is fully outside of its parent's clipping bounds...
+	if (maskedPageBounds === undefined) return true
+
+	// if the shape is fully outside of the viewport page bounds...
+	return !viewportPageBounds.includes(maskedPageBounds)
+}
+
+/**
+ * Incremental derivation of not visible shapes.
+ * Non visible shapes are shapes outside of the viewport page bounds and shapes outside of parent's clipping bounds.
+ *
+ * @param editor - Instance of the tldraw Editor.
+ * @returns Incremental derivation of non visible shapes.
+ */
+export const notVisibleShapes = (editor: Editor) => {
+	const isCullingOffScreenShapes = Number.isFinite(editor.renderingBoundsMargin)
+	const shapeHistory = editor.store.query.filterHistory('shape')
+	let lastPageId: TLPageId | null = null
+	let prevViewportPageBounds: Box
+
+	function fromScratch(editor: Editor): Set<TLShapeId> {
+		const shapes = editor.getCurrentPageShapeIds()
+		lastPageId = editor.getCurrentPageId()
+		const viewportPageBounds = editor.getViewportPageBounds()
+		prevViewportPageBounds = viewportPageBounds.clone()
+		const notVisibleShapes = new Set<TLShapeId>()
+		shapes.forEach((id) => {
+			if (isShapeNotVisible(editor, id, viewportPageBounds)) {
+				notVisibleShapes.add(id)
+			}
+		})
+		return notVisibleShapes
+	}
+	return computed<Set<TLShapeId>>('getCulledShapes', (prevValue, lastComputedEpoch) => {
+		if (!isCullingOffScreenShapes) return new Set<TLShapeId>()
+
+		if (isUninitialized(prevValue)) {
+			return fromScratch(editor)
+		}
+		const diff = shapeHistory.getDiffSince(lastComputedEpoch)
+
+		if (diff === RESET_VALUE) {
+			return fromScratch(editor)
+		}
+
+		const currentPageId = editor.getCurrentPageId()
+		if (lastPageId !== currentPageId) {
+			return fromScratch(editor)
+		}
+		const viewportPageBounds = editor.getViewportPageBounds()
+		if (!prevViewportPageBounds || !viewportPageBounds.equals(prevViewportPageBounds)) {
+			return fromScratch(editor)
+		}
+		const nextValue = new Set(prevValue)
+		let isDirty = false
```

Comment:
I prefer to lazily init the next value if possible
```suggestion
		let nextValue = null as null | Set<TLShapeId>
    const add = (id: TLShapeId) => {
      if (!nextValue) nextValue = new Set(prevValue)
      nextValue.add(id)
    }
    const delete = (id: TLShapeId) => {
      if (!nextValue) nextValue = new Set(prevValue)
      nextValue.delete(id)
    }
```
And then at the end you just do

```diff
return nextValue ?? prevValue
```

We have a helper for doing this kind of thing called `IncrementalSetConstructor` but it also keeps track of the diff between versions which we don't need to do here yet.

Reply:
Nice! Also moved some checks into these two new functions, it nicely simplified code.